igvm/vm.py

```diff
@@ -849,3 +866,81 @@ def vm_performance_value(self) -> float:
         ) * hv_perffactor_src
 
         return float(vm_performance_value)
+
+    def aws_get_instances_overview(
+            self, timeout: int = 5) -> Union[List, None]:
+        """AWS Get Instances Overview
+
+        Load or download the latest instances.json, which contains
+        a complete overview about all instance_types, their configuration,
+        performance and pricing.
+
+        :param: timeout: Timeout value for the get request
+
+        :return: VM types overview as list
+                 or None, if load/download failed
+        """
+
+        url = AWS_INSTANCES_OVERVIEW_URL
+        file = Path(AWS_INSTANCES_OVERVIEW_FILE)
+
+        if file.exists():
+            with open(AWS_INSTANCES_OVERVIEW_FILE) as f:
+                return json.load(f)
+        else:
+            r = urlopen(url, timeout=timeout)
+            if r.code == 200:
+                return json.loads(r.read())
+
+        return None
+
+    def aws_get_fitting_vm_types(self, overview: Union[List, None]) -> List:
+        """AWS Get Fitting VM types
+
+        Use the performance_value of the VM and multiply it with a static
+        factor to get the targeted ECU of AWS +-25% (which is kind of a
+        performance_value). Crawl the instances overview for fitting
+        vm_types for this targeted ECU range.
+
+        :param: overview: instances.json with a complete
+                          instance_types overview
+
+        :return: Fitting VM types as list
+        """
+
+        vm_performance_value = self.vm_performance_value()
+        region = self.dataset_obj['aws_placement'][:-1]
+
+        ecu_targeted = {
+            'min':
+                (vm_performance_value * AWS_ECU_FACTOR) -
+                (vm_performance_value * AWS_ECU_FACTOR * 0.25),
+            'max':
+                (vm_performance_value * AWS_ECU_FACTOR) +
+                (vm_performance_value * AWS_ECU_FACTOR * 0.25)
+        }
+
+        nested_dict = (
```

Comment:
what exactly do you want to achieve with this construct? i dont quite get it

Reply:
Using a `defaultdict()` is not sufficient here:
```
>>> vm_types = defaultdict()
>>> vm_types['foo']['foo2'] = 'bar'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'foo'
>>> 
```

I want to be able to create a nested dictionary, with the instance_type name as the key with multiple sub-keys and their corresponding values
```
>>> nested_dict = lambda: defaultdict(nested_dict)
>>> vm_types = nested_dict()
>>> vm_types['foo']['foo2'] = 'bar'
>>> vm_types
defaultdict(<function <lambda> at 0x7f064c426310>, {'foo': defaultdict(<function <lambda> at 0x7f064c426310>, {'foo2': 'bar'})})
>>> vm_types['foo']['foo2']
'bar'
```

It would look like this:
`defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'m5.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'm5.2xlarge', 'memory': 32.0, 'ECU': 37.0, 'price': 0.46}), 'i2.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'i2.2xlarge', 'memory': 61.0, 'ECU': 27.0, 'price': 2.026}), 'i3en.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'i3en.2xlarge', 'memory': 64.0, 'ECU': 37.0, 'price': 1.08}), 'p3.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'p3.2xlarge', 'memory': 61.0, 'ECU': 31.0, 'price': 3.823}), 'r5.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'r5.2xlarge', 'memory': 64.0, 'ECU': 37.0, 'price': 0.608}), 'm5d.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'm5d.2xlarge', 'memory': 32.0, 'ECU': 37.0, 'price': 0.544}), 'c5n.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'c5n.2xlarge', 'memory': 21.0, 'ECU': 39.0, 'price': 0.492}), 'c5.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'c5.2xlarge', 'memory': 16.0, 'ECU': 39.0, 'price': 0.388}), 'c5d.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'c5d.2xlarge', 'memory': 16.0, 'ECU': 39.0, 'price': 0.444}), 'c3.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'c3.2xlarge', 'memory': 15.0, 'ECU': 28.0, 'price': 0.516}), 'r4.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'r4.2xlarge', 'memory': 61.0, 'ECU': 31.0, 'price': 0.6402}), 'd2.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'd2.2xlarge', 'memory': 61.0, 'ECU': 28.0, 'price': 1.588}), 'c4.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'c4.2xlarge', 'memory': 15.0, 'ECU': 31.0, 'price': 0.454}), 'r5d.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'r5d.2xlarge', 'memory': 64.0, 'ECU': 37.0, 'price': 0.692}), 'i3.2xlarge': defaultdict(<function VM.aws_get_fitting_vm_types.<locals>.<lambda> at 0x7fc7d04130d0>, {'instance_type': 'i3.2xlarge', 'memory': 61.0, 'ECU': 31.0, 'price': 0.744})})`

Reply:
ah i see. i think this is a very complicated alias to just doing this:
```
vm_types = dict()

for t in overview:
    vm_types[t['instance_type']] = dict()
    vm_types[t['instance_type']]['instance_type'] = t['instance_type']
    vm_types[t['instance_type']]['memory'] = t['memory']
   ...
```
at least i would find that way more readable and clear

Reply:
Absolutely right, got rid of it. Thanks!